controllers/cesregistry/dogu_registrator.go

```diff
@@ -97,53 +98,69 @@ func (c *CesDoguRegistrator) UnregisterDogu(dogu string) error {
 	return nil
 }
 
-func (c *CesDoguRegistrator) addDoguToRegistry(dogu *core.Dogu) error {
-	err := c.doguRegistry.Register(dogu)
+func (c *CesDoguRegistrator) enableDoguInRegistry(dogu *core.Dogu) error {
+	err := c.doguRegistry.Enable(dogu)
 	if err != nil {
-		return fmt.Errorf("failed to register dogu %s: %w", dogu.GetSimpleName(), err)
+		return fmt.Errorf("failed to enable dogu: %w", err)
 	}
+	return nil
+}
 
-	err = c.doguRegistry.Enable(dogu)
+func (c *CesDoguRegistrator) registerDoguInRegistry(dogu *core.Dogu) error {
+	err := c.doguRegistry.Register(dogu)
 	if err != nil {
-		return fmt.Errorf("failed to enable dogu: %w", err)
+		return fmt.Errorf("failed to register dogu %s: %w", dogu.GetSimpleName(), err)
 	}
-
 	return nil
 }
 
-// registerNewKeys creates a new key pair and registers it with the dogu in the registry. Any pre-existing keys will
-// then no longer work.
+// registerNewKeys creates a new key pair and registers it with the dogu in the registry. If the private key exists it
+// will be used to generate a new public key.
 func (c *CesDoguRegistrator) registerNewKeys(ctx context.Context, doguResource *k8sv1.Dogu, dogu *core.Dogu) error {
-	keyPair, err := c.createKeypair()
+	keyProvider, err := resource.GetKeyProvider(c.registry)
 	if err != nil {
-		return fmt.Errorf("failed to create keypair: %w", err)
+		return err
 	}
 
-	err = c.writePublicKey(keyPair.Public(), dogu)
+	secret, err := doguResource.GetPrivateKeySecret(ctx, c.client)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return err
+	}
+
+	if apierrors.IsNotFound(err) {
+		keyPair, err := keyProvider.Generate()
+		if err != nil {
+			return fmt.Errorf("failed to generate key pair: %w", err)
+		}
+
+		return c.writeKeyPair(ctx, doguResource, dogu, keyPair)
+	}
```

Comment:
The granularity is way different from the methods directly above this function, even though you already describe the scenarios in the doc comments. Why did you not extracted code into methods like `createKeyPair()` and `recreatePubKey()` or anything similar?